### PR TITLE
Akka Coordination Lease for Kubernetes API server

### PR DESCRIPTION
**Pull request by chbatey**

### CODE REVIEW

File: docs/src/main/paradox/kubernetes-lease.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+@@@
+
+This module is an implementation of an [Akka Coordination Lease](https://doc.akka.io/docs/akka/current/coordination.html#lease) backed 
+by a [Custom Resource Definition (CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) in Kubernetes.
+Resources in Kubernetes offer [concurrency control and consistency](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) 
+that have been used to build a distributed lease/lock.
+
+A lease can be used for:
+
+* @extref[Split Brain Resolver](akka:split-brin-resolver.html). An additional safety measure so that only one SBR instance can make the decision to remain up.
```
Comment (patriknw) [inline docs/src/main/paradox/kubernetes-lease.md:17]: typo: brin

File: docs/src/main/paradox/kubernetes-lease.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+* @extref:[Cluster Singleton](akka:cluster-singleton.html#lease). A singleton manager can be configured to acquire a lease before creating the singleton.
```
Comment (patriknw) [inline docs/src/main/paradox/kubernetes-lease.md:18]: do these links end up in the right place? `typed/cluster-singleton.html` ?

File: docs/src/main/paradox/kubernetes-lease.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+* @extref:[Cluster Sharding](akka:cluster-sharding.html#lease). Each `Shard` can be configured to acquire a lease before creating entity actors.
+
+In all cases the use of the lease increases the consistency of the feature. However, as the Kubernetes API server 
+and its backing `etcd` cluster can also be subject to failure and network issues any use of this lease can reduce availability. 
+
+### Lease Instances
+
+* With @extref[Split Brain Resolver](akka:split-brin-resolver.html) there will be one lease per Akka Cluster
```
Comment (patriknw) [inline docs/src/main/paradox/kubernetes-lease.md:26]: typo

File: build.sbt [hunk 4]
```diff
@@ -111,6 +113,43 @@ lazy val `cluster-bootstrap` = project
   )
   .dependsOn(`akka-management`)
 
+lazy val `lease-kubernetes` = project
```
Comment (patriknw) [inline build.sbt:116]: include in some aggregate?

File: build.sbt [hunk 5]
```diff
@@ -111,6 +113,43 @@ lazy val `cluster-bootstrap` = project
...
+  .in(file("lease-kubernetes"))
+  .enablePlugins(AutomateHeaderPlugin)
+  .settings(
+    name := "akka-lease-kubernetes",
```
Comment (patriknw) [inline build.sbt:120]: organization is by default `organization := "com.lightbend.akka.management"`, I guess that works for this

Comment (chbatey) [reply build.sbt:120]: bit weird to have management in the name but would also be weird to have a project in this repo without management in the name

File: docs/src/main/paradox/kubernetes-lease.md (hunk 2 above)
Comment (chbatey) [reply docs/src/main/paradox/kubernetes-lease.md:18]: Ah they go to 2.5! Will fix that

File: build.sbt (hunk 5 above)
Comment (patriknw) [reply build.sbt:122]: On the other hand akka-coordination is in com.typesafe.akka and we have used com.lightbend.akka or com.typesafe.akka also for things located outside the akka/akka repo.

Not sure, but we could use `com.lightbend.akka`.

Comment (patriknw) [reply build.sbt:122]: bootstrap is using `com.lightbend.akka.management`. Maybe we'll just stick to that instead of adding more?

Comment (chbatey) [reply build.sbt:122]: Ah sorry it was the discovery impls that didn't use management in the name e.g. `com.lightbend.akka.discovery`

Yes let's just stick to com.lightbend.akka.management and be done with it!

File: lease-kubernetes/src/main/resources/reference.conf [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    # Namespace to create the lock in. If set to something other than "<namespace>" then overrides any value
+    # in "namespace-path"
+    namespace = "<namespace>"
+
+    # How often to write time into CRD so that if the holder crashes
+    # another node can take the lease after a given timeout. If left blank then the default is
+    # max(5s, heartbeat-timeout / 10)
+    heartbeat-interval = ""
+    #heartbeat-interval = 12s
```
Comment (johanandren) [inline lease-kubernetes/src/main/resources/reference.conf:31]: A bit confusing, perhaps just mention that the default will end up being 12s in the comment instead?

File: lease-kubernetes/src/main/resources/reference.conf [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    # How long a lease must not be updated before another node can assume
+    # the holder has crashed.
+    # If the lease holder hasn't crashed its next heart beat will fail due to the version
+    # having been updated
+    heartbeat-timeout = 120s
+
+    # The individual timeout for each HTTP request. Defaults to 2/5 of the lease-operation-timoeut
+    # Can't be greater than then lease-operation-timeout
+    api-server-request-timeout = ""
+    #api-server-request-timeout = 2s
```
Comment (johanandren) [inline lease-kubernetes/src/main/resources/reference.conf:42]: Same here, or is this some convention we do that I have missed?

File: lease-kubernetes/src/main/scala/akka/coordination/lease/kubernetes/KubernetesLease.scala [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  override def checkLease(): Boolean = leaseTaken.get()
+
+  // TODO (followup) make release idempotent so client can retry?
+  override def release(): Future[Boolean] = {
+    // replace with transform once 2.11 dropped
+    (leaseActor ? Release())
+      .flatMap {
+        case LeaseReleased       => Future.successful(true)
+        case InvalidRequest(msg) => Future.failed(new LeaseException(msg))
+      }
```
Comment (johanandren) [inline lease-kubernetes/src/main/scala/akka/coordination/lease/kubernetes/KubernetesLease.scala:51]: Could use parasitic/samethread ec

File: lease-kubernetes/src/main/scala/akka/coordination/lease/kubernetes/KubernetesLease.scala [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      .flatMap {
+        case LeaseAcquired       => Future.successful(true)
+        case LeaseTaken          => Future.successful(false)
+        case InvalidRequest(msg) => Future.failed(new LeaseException(msg))
+      }
+      .recoverWith {
+        case _: AskTimeoutException =>
+          Future.failed[Boolean](new LeaseTimeoutException(
+            s"Timed out trying to acquire lease [${settings.leaseName}, ${settings.ownerName}]. It may still be taken."))
+      }
```
Comment (johanandren) [inline lease-kubernetes/src/main/scala/akka/coordination/lease/kubernetes/KubernetesLease.scala:75]: Could use parasitic/samethread ec

File: lease-kubernetes/src/main/scala/akka/coordination/lease/kubernetes/internal/KubernetesApiImpl.scala [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+import akka.coordination.lease.{ LeaseException, LeaseTimeoutException }
+import akka.stream.ActorMaterializer
+import com.typesafe.sslconfig.akka.AkkaSSLConfig
+import com.typesafe.sslconfig.ssl.TrustStoreConfig
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+// TODO Extension so a singleton + limit the number of concurrent requests to the API server
```
Comment (johanandren) [inline lease-kubernetes/src/main/scala/akka/coordination/lease/kubernetes/internal/KubernetesApiImpl.scala:29]: issue or fix

File: lease-kubernetes/README.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
+# Kubernetes Lease
```
[the remark this answers is not in the document]
Comment (chbatey) [reply lease-kubernetes/README.md:1]: Yep, will remove apart from the last bit about internals

File: lease-kubernetes/src/main/resources/reference.conf (hunk 6 above)
Comment (chbatey) [reply lease-kubernetes/src/main/resources/reference.conf:31]: 👍

**Pull request merged**